<div class="alert alert-info" style="background-color:#001b79; color:white; padding:0px 10px; border-radius:5px;"><h2 style='margin:10px 5px'>ADA BOOST CLASSIFIER </h2>
</div>

### Boosting:	

- Concept of converting a group of week learner to strong learners is called Boosting.
- The term ‘Boosting’ refers to a family of algorithms which converts weak learner to strong learners.


- AdaBoost (Adaptive Boosting) :		
		
>  It fits a sequence of weak learners on different weighted training data.	

> It starts by predicting original data set and gives equal weight to each observation.

> If prediction is incorrect using the first learner, then it gives higher weight to observation which have been predicted incorrectly. 	

> Being an iterative process, it continues to add learner(s) until a limit is reached in the number of models or accuracy.	
    
> We can use AdaBoost algorithms for both classification and regression problem.


#### n_estimators: It controls the number of weak learners.

#### learning_rate:Controls the contribution of weak learners in the final combination. There is a trade-off between learning_rate and n_estimators.

#### base_estimators: It helps to specify different ML algorithm.


<div class="alert alert-info" style="background-color:#001b32; color:white; padding:0px 10px; border-radius:5px;"><h2 style='margin:10px 5px'>Problem  Statement </h2>
</div>

## Module 14 Ada Boost classifier

### Data set:
- This is diabetic patient data, class is the o/p value

##  1. Import all Librariy

In [25]:
import pandas as pd
from sklearn import model_selection

from sklearn.ensemble import AdaBoostClassifier

In [26]:
# Read data, no header in txt, so added header=None
df = pd.read_csv(r'..\Dataset\Diabetes.txt', 
                 sep=",", header=None)  # Data having comma,  so removed that

# Dataset dont have column name, that we have separately. so added here
df.columns = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
array = df.values
X = array[:,0:8]  # Col preg to ge
Y = array[:,8]  # Col class
df.head()

,preg,plas,pres,skin,test,mass,pedi,age,class
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


##  2. Apply ada boost classifier and then use Cross validation to see accuracy

In [27]:
# This is 10 fold cross validation step
kfold = model_selection.KFold(n_splits=10, random_state=7)

# Ada boost classifier assign
model = AdaBoostClassifier(n_estimators=30, random_state=7)

score_ada = model_selection.cross_val_score(model, X, Y, cv=kfold).mean() 

## Apply Cross validation on adaboost classifier model, with k-fold/cv=10, and take mean which is the avg of all 10 accuracy
results = model_selection.cross_val_score(model, X, Y, cv=kfold)
print(results.mean())

0.760457963089542


## 3.  Apply normal knn without adaboost

In [9]:
from sklearn.neighbors import KNeighborsClassifier

#KNN model assign

knn = KNeighborsClassifier(n_neighbors=4)

score_knn = model_selection.cross_val_score(knn, X, Y, cv=kfold).mean() 
## Apply Cross validation on logreg model, with k-fold/cv=10, and take mean which is the avg of all 10 accuracy
print (model_selection.cross_val_score(knn, X, Y, cv=kfold).mean())

0.7212747778537253


## 4. Improvement in %

In [24]:
print ("Improvement in %" ,((score_ada - score_knn)*100/score_knn))

Improvement in % 5.4324907010353165


## 5. Conclusion:
- With adabooster classifier accuracy score is more(76%) compared to just direct knn(72%)
- By this we can see adabooster improved week learner to strong learner